File: other_problems/5_closure_test_1.rb
def memorizer(memo, formula)
  lambda = lambda do |n|
    result = memo[n]

    if result.nil?
      result = formula.call(lambda, n)
      memo[n] = result
    end
    return result
  end

  return lambda
end

fibonacci = memorizer([0, 1], lambda { |recur, n| return recur.call(n-1) + recur.call(n-2) })
factorial = memorizer([1, 1], lambda { |recur, n| return recur.call(n-1) * n})

p fibonacci.call(10)
p factorial.call(3)

File: other_problems/5_closure_test_1.js
var memorizer = function (memo, formula) {
  var recur = function(n) {
    var result = memo[n];
    if (typeof result !== 'number') {
      result = formula(recur, n);
      memo[n] = result;
    }
    return result;
  }

  return  recur;
};

var fibonacci = memorizer([0, 1], function(recur, n) {
  return recur(n-1) + recur(n-2);
});

var factorial = memorizer([1, 1], function(recur, n) {
  return n * recur(n-1);
});

console.log(fibonacci(10));
console.log(factorial(3));
